### ISSUE TITLE
Rule proposal: Disallow number literals where enums are expected

### ISSUE THREAD

**Issue by mohsen1** [typescript-eslint/typescript-eslint#308]
Do not allow number literals to be used instead of enums. I could think of three places that this can happen:

* Function arguments 
* Binary expressions
* Assignment expressions

```ts
enum Foo {
    ONE,
    TWO
}

function getFoo(f: Foo) {

    if (f === Foo.ONE) { // Good

    }

    if (f === 1) { // Bad

    }

    let ff: Foo;

    ff = Foo.TWO // Good
    ff = 1 // Bad

}

getFoo(Foo.ONE) // Good
getFoo(1) // Bad
```

### Rule name
I need help with naming it according to conventions 

### Options 
I don't think this rule needs any options. I don't see why anyone wants to use number literals for enums.

**Comment by bradzacher** [comment]
This could be expanded to just "disallow non-enum comparisons when using enum types"

So that it covers all of the cases (which are valid typescript, but error-prone):
```ts
enum FooImplicit {
    A,
    B,
}
FooImplicit.A === 0

enum FooNum {
    A = 1,
    B = 2,
}
FooNum.A === 1

enum FooStr {
    A = 'a',
    B = 'b',
}
FooStr.A === 'a'
```

**Comment by bradzacher** [comment]
TS 4.3 brings in some checks here that solves this for number enums!
[Union Enums Cannot Be Compared to Arbitrary Numbers](https://devblogs.microsoft.com/typescript/announcing-typescript-4-3-rc/#union-enums-cannot-be-compared-to-arbitrary-numbers)

**Comment by ChocolateLoverRaj** [comment]
@bradzacher @JoshuaKGoldberg if someone makes a PR for this feature, what should the name of the rule be?

**Comment by bradzacher** [comment]
It's really hard to name this. Maybe something like `no-misused-enums`?

Given this rule will cover assignments and comparisons I can't think of a good name which covers both cases succinctly.

**Comment by JoshuaKGoldberg** [comment]
@ChocolateLoverRaj are you interested in & have time for sending a PR?

No worries if you don't, I can send one soon if not!

**Comment by ChocolateLoverRaj** [comment]
> @ChocolateLoverRaj are you interested in & have time for sending a PR?
> 
> No worries if you don't, I can send one soon if not!

Yes, i can make a PR. it's my first time doing an eslint related PR though.

**Comment by Zamiell** [comment]
@ChocolateLoverRaj Any updates on this?

**Comment by Zamiell** [comment]
> It's really hard to name this. Maybe something like no-misused-enums?

How about `strict-enum-comparisons`?

**Comment by Zamiell** [comment]
Also, I would want greater-than and less-than comparison operators to be illegal:

```ts
// bad
if (Foo.A > Foo.B) {}
```

**Comment by Zamiell** [comment]
I'm trying my hand at this, but I'm stuck at comparing the types.

```ts
const variableType = typeChecker.getTypeAtLocation(typescriptNode);
const initializerType = typeChecker.getTypeAtLocation(typescriptNode.initializer);
const areTypesEqual = variableType === initializerType; // Prints false
```

What is the correct way to see if the types are equal? I'm looking for an ID field or something, but I don't see anything like that. In the helper functions, I do see `getTypeName`, but its possible for two different enums to exist with the same name, right? So I presume that comparing by using the name would not be safe.

**Comment by Zamiell** [comment]
I proceeded with just using the name to compare for now.
I did more work on this today.
Everything works great, except for detecting invalid function calls, like:
```ts
function fruitFunction(fruit: Fruit) {}
fruitFunction(0);
```

I'm stuck on how to get the type of the matching function argument:

```ts
CallExpression(node): void {
  const identifier = node.callee;
  const functionType = getTypeFromNode(identifier);

  for (let i = 0; i < node.arguments.length; i++) {
    const expressionArgument = node.arguments[i];
    const expressionArgumentType = getTypeFromNode(argument);
    const expressionArgumentTypeName = getTypeName(argumentType);

    const functionArgumentType = getFunctionArgumentType(functionType, i);
    const functionArgumentTypeName = getTypeName(functionArgumentType);
    
    if (expressionArgumentTypeName  !== functionArgumentTypeName) {
      context.report({ node, messageId: 'incorrectFunctionArgument' });
    }
  }
}
```
```ts
function getFunctionArgumentType(functionType: ts.Type, i: number) {
  // ???????????????????
}
```

**Comment by Zamiell** [comment]
I made more progress today. It turns out the trick to get function argument types was to use the `getCallSignaturesOfType` helper function.

Now I'm stuck on a new problem. I don't know how to use ESLint and/or the TypeScript API to get the constituent types of a union type. Until I can figure this out, the rule doesn't catch things like:

```ts
const fruit: Fruit | null = 0;
```

**Comment by JoshuaKGoldberg** [comment]
> > It's really hard to name this. Maybe something like no-misused-enums?
> 
> How about `strict-enums`?

That's a bit catch-all, maybe... `strict-enum-comparisons`?

> comparing the types

:sob: https://github.com/microsoft/TypeScript/issues/9879 strikes yet again.

> way to see if the types are equal

You could try looking at the declarations with `symbol.getDeclarations()`. ✨ 

> `getCallSignaturesOfType`

For the curious, that's a small wrapper around `type.getCallSignatures()` that also accounts for unions and intersections: https://github.com/ajafff/tsutils/blob/03b4df1/util/type.ts#L121

> constituent types of a union

`type.isUnion()` is type `(): this is ts.UnionType` -- which gives a hint that `ts.UnionType` has what you need.

```ts
if (type.isUnion()) {
  type.types; // UnionOrIntersectionType.types: ts.Type[]
}
```

---

Looks like you're making a lot of great progress. Really exciting!

**Comment by Zamiell** [comment]
> That's a bit catch-all, maybe... strict-enum-comparisons?

Currently, the rule does more than look at comparisons. It looks at:

1) Declarations
2) Assignments
3) Incrementing/Decrementing
4) Comparisons
5) Function arguments

You can look through the strict-enums.test.ts to see.
It's at 1184 lines currently and growing.

**Comment by Zamiell** [comment]
> You could try looking at the declarations with symbol.getDeclarations()

I actually found you that you can just use triple equals to compare types.
My initial tests with triple equals didn't work because I was comparing `Fruit` to `Fruit.Apple`, which are different types, so I got confused and burned a bunch of hours trying other things.

**Comment by Zamiell** [comment]
> if (type.isUnion()) {

Yup, that works great, thanks.

**Comment by JoshuaKGoldberg** [comment]
Ah, we have an existing `no-misused-promises` rule that also does a lot more than comparisons. Maybe that's a good naming precedent?

**Comment by Zamiell** [comment]
Well, there's also an existing rule called `strict-boolean-expressions`.

**Comment by JoshuaKGoldberg** [comment]
![Homer Simpson swinging on a wrecking ball between a giant rock and a bar named "A Hard Place"](https://user-images.githubusercontent.com/3335181/165420568-b6e279c2-778f-4e90-a217-0cf7def59622.gif)

**Comment by bradzacher** [comment]
For navigating unions you can use `tsutils.unionTypeParts(type)` which will handle both the union and non-union cases automatically.

---

For comparing two types - you can do an equality check. If the type is the *same* type, then TS will return a referentially equal object.

For enums because they are a "primitive" type (non object), TS will do the hard work of resolving past aliases and such for you.

EG for the following code, the resolved types (via `checker.getTypeAtLocation(node)`) for the variables will be the same type object.
```ts
enum Foo { A }
declare const x: Foo.A;
type T = Foo.A;
declare const y: T;
```

You can play around in https://ts-ast-viewer.com/ to validate this assumption.

If we were dealing with object types - things would get much harder, but enums are simple!

---

Re: naming - maybe `no-misused-enums` is a good name?
Using `-comparisons` isn't great because ideally we should also assignments as well as comparisons:
```ts
enum Foo { A }

const x: Foo = 0; // should error
if (x === 0) {} // should error
```

**Comment by Zamiell** [comment]
Also note that `strict-enums` follows the TypeScript compiler flag convention of `--strict`, `--strictBindCallApply`, `--strictFunctionTypes`, `--strictNullChecks`, and so on.

If this code was ever submitted as a pull request to TypeScript itself, it would almost certainly be named `--strictEnums`.

**Comment by Zamiell** [comment]
Ok, I did more work today, and all the tests are passing now.
It uses using strict equality to compare the types, and uses the `unionTypeParts` helper function to decompose unions.
I'll mark the PR as not being a draft anymore.

**Comment by Zamiell** [comment]
The lint rule uncovers some bugs in the typescript-eslint codebase.
Thus, linting in CI fails.

Should I fix all of the bugs as part of this PR or should that be done in a separate PR?

**Comment by JoshuaKGoldberg** [comment]
Could you fix them in the PR please? If you don't have time a TODO comment would suffice. In the past we've found the typescript-eslint codebase to be surprisingly good at exposing some edge cases in PRs.

**Comment by JoshuaKGoldberg** [comment]
From #6107: I don't have the time to work on this either, so if anybody else wants to pick it up, please do!

**Comment by Zamiell** [comment]
If anyone wants it, I implemented the rule here as part of `eslint-plugin-isaacscript`: https://github.com/IsaacScript/isaacscript/blob/main/packages/eslint-plugin-isaacscript/docs/rules/strict-enums.md

**Comment by loicraux** [comment]
Would be nice if it could also handle switch statements... No errors are reported for instance with this code where the `switch` statement deals with a string and compare against an string-based enum `EventKeys`:
```ts
    @bound private handleKeydown(keyboardEvent: KeyboardEvent): void {
        let doStopPropagationAndPreventDefault = false;
        assertIsDefined(this.controlledPopupMenu);
        switch (keyboardEvent.key) {
            case EventKeys.SPACE:
            case EventKeys.ENTER:
            case EventKeys.DOWN:
                this.controlledPopupMenu.open();
                this.controlledPopupMenu.setFocusToFirstVisibleItem();
                doStopPropagationAndPreventDefault = true;
                break;

            case EventKeys.UP:
                this.controlledPopupMenu.open();
                this.controlledPopupMenu.setFocusToLastVisibleItem();
                doStopPropagationAndPreventDefault = true;
                break;

            default:
                break;
        }

        if (doStopPropagationAndPreventDefault) {
            keyboardEvent.stopPropagation();
            keyboardEvent.preventDefault();
        }
    }
```

**Comment by Zamiell** [comment]
@loicraux I've since stripped out the code in [`complete/strict-enums`](https://github.com/complete-ts/complete/blob/main/packages/eslint-plugin-complete/docs/rules/strict-enums.md) that overlaps with the official [`no-unsafe-enum-comparison`](https://typescript-eslint.io/rules/no-unsafe-enum-comparison) rule.

But yes, what you posted is a bug, let's track that in #7509.

**Comment by RedGuy12** [comment]
I think this should be closed now?

**Comment by Zamiell** [comment]
@RedGuy12 It cannot be closed because the rule only checks for comparisons and doesn't check for assignment.

e.g.

```ts
enum Fruit {
  Apple,
  Banana,
}

let fruit = Fruit.Apple;
fruit = 1; // No error!
```

If you want to prevent this unsafe pattern in your code, use the [`complete/strict-enums`](https://github.com/complete-ts/complete/blob/main/packages/eslint-plugin-complete/docs/rules/strict-enums.md) rule.

**Comment by ThFischer** [comment]
Even worse:
```typescript
enum FooBar {
    foo,
    bar
}
const foo: FooBar = NaN; // No error
const bar: FooBar = 2;   // Type '2' is not assignable to type 'FooBar'
```

**Comment by kirkwaiblinger** [comment]
> Even worse:
> 
> enum FooBar {
>     foo,
>     bar
> }
> const foo: FooBar = NaN; // No error
> const bar: FooBar = 2;   // Type '2' is not assignable to type 'FooBar'

_Technically_, `NaN` isn't a numeric literal; it's just a variable with type `number`. So I guess this code would fall under unsafe `number` -> numeric enum assignment, same as 

```ts
enum FooBar {
    foo,
    bar
}


function f(x: number): FooBar {
  return x;
}
```

Suprisingly, TS allows this. I thought it was supposed to be fixed in https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#all-enums-are-union-enums tbh but I guess not?

**Comment by bradzacher** [comment]
TS intentionally does not protect numeric enums with any checks because "bit flags enums" are a common usecase and those require arithmetic (such as binary `|` or's) which always result in `number`.

It's a known point of unsafety from this style of code.
```ts
enum Flags {
  A = 1 << 0,
  B = 1 << 1,
  C = 1 << 2,
}

// intended usecase
const valid: Flags = Flags.A | Flags.B;
const is_A_set = valid & Flags.A === Flags.A;

// unfortunately it also allows unsafe crap too
const wtf: Flags = 99 + 1;
const seriously: Flags = Number.MAX_SAFE_INTEGER;
```
[playground](https://www.typescriptlang.org/play/?#code/KYOwrgtgBAYgNgQwOYGcoG8BQUoEEoC8UAjFADxlQAMANNlAEKEnmXF04DCzpFUATHQC+mTAHoxUAJYgALqAAmwBVDApgAYwTrMGgPYgUsqADcEcKQoBcsRKmbxkKAHT4APrafOGAbl0GjaRQAfVxg9WMiMwsVADJPVFdCAiJHRNw-cUkwEAAzPQAnWRyEeTgAT2ljcxQ9KHM4PQB3NByUBFzgKA0ChAAHKFk9PX9DYybZXJs0tCIATjmoAGoSP30xqHUCqT01Cum7WagAOUgAI2AC5wBZXAANYIBlXBgAUWCASWOAFVeAcVeACU-EA)
